Error when running STS4 Tooling

**Description:**
This seems like the Runtime error which is lack of some range checking
```
SEVERE: Internal error: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
java.util.concurrent.CompletionException: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1629)
	at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1618)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at org.wso2.lsp4intellij.client.DynamicRegistrationMethods.forName(DynamicRegistrationMethods.java:44)
	at org.wso2.lsp4intellij.client.LanguageClientImpl.lambda$null$1(LanguageClientImpl.java:76)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at org.wso2.lsp4intellij.client.LanguageClientImpl.lambda$registerCapability$2(LanguageClientImpl.java:74)
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1626)
	... 5 more
```

--- a/src/main/java/org/wso2/lsp4intellij/IntellijLanguageClient.java
+++ b/src/main/java/org/wso2/lsp4intellij/IntellijLanguageClient.java
@@ -29,18 +29,6 @@ import com.intellij.openapi.project.ProjectManager;
 import com.intellij.openapi.util.Disposer;
 import com.intellij.openapi.vfs.VirtualFile;
 import com.intellij.openapi.vfs.VirtualFileManager;
-import java.util.AbstractMap;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
-import java.util.stream.Stream;
 import org.apache.commons.lang3.tuple.ImmutablePair;
 import org.apache.commons.lang3.tuple.MutablePair;
 import org.apache.commons.lang3.tuple.Pair;
@@ -56,14 +44,27 @@ import org.wso2.lsp4intellij.requests.Timeouts;
 import org.wso2.lsp4intellij.utils.ApplicationUtils;
 import org.wso2.lsp4intellij.utils.FileUtils;
 
+import java.util.AbstractMap;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
 public class IntellijLanguageClient implements ApplicationComponent {
+
+    private static Logger LOG = Logger.getInstance(IntellijLanguageClient.class);
 
     private static final Map<Pair<String, String>, LanguageServerWrapper> extToLanguageWrapper = new ConcurrentHashMap<>();
     private static Map<String, Set<LanguageServerWrapper>> projectToLanguageWrappers = new ConcurrentHashMap<>();
     private static Map<String, LanguageServerDefinition> extToServerDefinition = new ConcurrentHashMap<>();
     private static Map<String, LSPExtensionManager> extToExtManager = new ConcurrentHashMap<>();
-
-    private static Logger LOG = Logger.getInstance(IntellijLanguageClient.class);
 
     @Override
     public void initComponent() {
@@ -141,14 +142,14 @@ public class IntellijLanguageClient implements ApplicationComponent {
     public static void editorOpened(Editor editor) {
         VirtualFile file = FileDocumentManager.getInstance().getFile(editor.getDocument());
         if (!FileUtils.isFileSupported(file)) {
-            LOG.debug("Handling open on a editor which host a virtual file");
+            LOG.debug("Handling open on a editor which host a LightVirtual/Null file");
             return;
         }
 
         Project project = editor.getProject();
         String rootPath = FileUtils.editorToProjectFolderPath(editor);
         String rootUri = FileUtils.pathToUri(rootPath);
-        if (file != null && rootUri != null) {
+        if (rootUri != null && project != null) {
             ApplicationUtils.pool(() -> {
                 String ext = file.getExtension();
                 final String fileName = file.getName();
@@ -185,14 +186,9 @@ public class IntellijLanguageClient implements ApplicationComponent {
 
                         // Update project mapping for language servers
                         final String projectUri = FileUtils.pathToUri(project.getBasePath());
-                        Set<LanguageServerWrapper> wrappers = projectToLanguageWrappers.get(projectUri);
-                        if (wrappers == null) {
-                            wrappers = new HashSet<>();
-                            projectToLanguageWrappers.put(projectUri, wrappers);
-                        }
-                        if (!wrappers.contains(wrapper)) {
-                            wrappers.add(wrapper);
-                        }
+                        Set<LanguageServerWrapper> wrappers = projectToLanguageWrappers
+                                .computeIfAbsent(projectUri, k -> new HashSet<>());
+                        wrappers.add(wrapper);
                     } else {
                         LOG.info("Wrapper already existing for " + ext + " , " + rootUri);
                     }
@@ -201,7 +197,12 @@ public class IntellijLanguageClient implements ApplicationComponent {
                 }
             });
         } else {
-            LOG.warn("File for editor " + editor.getDocument().getText() + " is null");
+            if (rootUri == null) {
+                LOG.warn("File for editor " + editor.getDocument().getText() + " is null");
+            }
+            if (project == null) {
+                LOG.warn("Project for editor " + editor.getDocument().getText() + " is null");
+            }
         }
     }
 
@@ -213,22 +214,17 @@ public class IntellijLanguageClient implements ApplicationComponent {
     public static void editorClosed(Editor editor) {
         VirtualFile file = FileDocumentManager.getInstance().getFile(editor.getDocument());
         if (!FileUtils.isFileSupported(file)) {
-            LOG.debug("Handling close on a editor which host a virtual file");
+            LOG.debug("Handling close on a editor which host a LightVirtual/Null file");
             return;
         }
 
-        if (file != null) {
-            ApplicationUtils.pool(() -> {
-                String ext = file.getExtension();
-                LanguageServerWrapper serverWrapper = LanguageServerWrapper.forEditor(editor);
-                if (serverWrapper != null) {
-                    LOG.info("Disconnecting " + FileUtils.editorToURIString(editor));
-                    serverWrapper.disconnect(editor);
-                }
-            });
-        } else {
-            LOG.warn("File for editor " + editor.getDocument().getText() + " is null");
-        }
+        ApplicationUtils.pool(() -> {
+            LanguageServerWrapper serverWrapper = LanguageServerWrapper.forEditor(editor);
+            if (serverWrapper != null) {
+                LOG.info("Disconnecting " + FileUtils.editorToURIString(editor));
+                serverWrapper.disconnect(editor);
+            }
+        });
     }
 
     /**

--- a/src/main/java/org/wso2/lsp4intellij/client/DefaultLanguageClient.java
+++ b/src/main/java/org/wso2/lsp4intellij/client/DefaultLanguageClient.java
@@ -39,15 +39,16 @@ import org.wso2.lsp4intellij.requests.WorkspaceEditHandler;
 import org.wso2.lsp4intellij.utils.ApplicationUtils;
 import org.wso2.lsp4intellij.utils.FileUtils;
 
+import javax.swing.Icon;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.FutureTask;
-import javax.swing.*;
 
 public class DefaultLanguageClient implements LanguageClient {
 
@@ -79,9 +80,9 @@ public class DefaultLanguageClient implements LanguageClient {
     public CompletableFuture<Void> registerCapability(RegistrationParams params) {
         return CompletableFuture.runAsync(() -> params.getRegistrations().forEach(r -> {
             String id = r.getId();
-            DynamicRegistrationMethods method = DynamicRegistrationMethods.forName(r.getMethod());
-            Object options = r.getRegisterOptions();
-            registrations.put(id, method);
+            Optional<DynamicRegistrationMethods> method = DynamicRegistrationMethods.forName(r.getMethod());
+            method.ifPresent(dynamicRegistrationMethods -> registrations.put(id, dynamicRegistrationMethods));
+
         }));
     }
 
@@ -89,7 +90,7 @@ public class DefaultLanguageClient implements LanguageClient {
     public CompletableFuture<Void> unregisterCapability(UnregistrationParams params) {
         return CompletableFuture.runAsync(() -> params.getUnregisterations().forEach((Unregistration r) -> {
             String id = r.getId();
-            DynamicRegistrationMethods method = DynamicRegistrationMethods.forName(r.getMethod());
+            Optional<DynamicRegistrationMethods> method = DynamicRegistrationMethods.forName(r.getMethod());
             if (registrations.containsKey(id)) {
                 registrations.remove(id);
             } else {
@@ -97,8 +98,8 @@ public class DefaultLanguageClient implements LanguageClient {
                 for (Map.Entry<String, DynamicRegistrationMethods> entry : registrations.entrySet()) {
                     inverted.put(entry.getValue(), entry.getKey());
                 }
-                if (inverted.containsKey(method)) {
-                    registrations.remove(inverted.get(method));
+                if (method.isPresent() && inverted.containsKey(method.get())) {
+                    registrations.remove(inverted.get(method.get()));
                 }
             }
         }));

--- a/src/main/java/org/wso2/lsp4intellij/client/DynamicRegistrationMethods.java
+++ b/src/main/java/org/wso2/lsp4intellij/client/DynamicRegistrationMethods.java
@@ -16,22 +16,32 @@
 package org.wso2.lsp4intellij.client;
 
 import java.util.Arrays;
-import java.util.stream.Collectors;
+import java.util.Optional;
 
 /**
  * Enum for methods which may support DynamicRegistration
  */
 public enum DynamicRegistrationMethods {
-    DID_CHANGE_CONFIGURATION("workspace/didChangeConfiguration"), DID_CHANGE_WATCHED_FILES(
-            "workspace/didChangeWatchedFiles"), SYMBOL("workspace/symbol"), EXECUTE_COMMAND(
-            "workspace/executeCommand"), SYNCHRONIZATION("textDocument/synchronization"), COMPLETION(
-            "textDocument/completion"), HOVER("textDocument/hover"), SIGNATURE_HELP(
-            "textDocument/signatureHelp"), REFERENCES("textDocument/references"), DOCUMENT_HIGHLIGHT(
-            "textDocument/documentHighlight"), DOCUMENT_SYMBOL("textDocument/documentSymbol"), FORMATTING(
-            "textDocument/formatting"), RANGE_FORMATTING("textDocument/rangeFormatting"), ONTYPE_FORMATTING(
-            "textDocument/onTypeFormatting"), DEFINITION("textDocument/definition"), CODE_ACTION(
-            "textDocument/codeAction"), CODE_LENS("textDocument/codeLens"), DOCUMENT_LINK(
-            "textDocument/documentLink"), RENAME("textDocument/rename");
+
+    DID_CHANGE_CONFIGURATION("workspace/didChangeConfiguration"),
+    DID_CHANGE_WATCHED_FILES("workspace/didChangeWatchedFiles"),
+    SYMBOL("workspace/symbol"),
+    EXECUTE_COMMAND("workspace/executeCommand"),
+    SYNCHRONIZATION("textDocument/synchronization"),
+    COMPLETION("textDocument/completion"),
+    HOVER("textDocument/hover"),
+    SIGNATURE_HELP("textDocument/signatureHelp"),
+    REFERENCES("textDocument/references"),
+    DOCUMENT_HIGHLIGHT("textDocument/documentHighlight"),
+    DOCUMENT_SYMBOL("textDocument/documentSymbol"),
+    FORMATTING("textDocument/formatting"),
+    RANGE_FORMATTING("textDocument/rangeFormatting"),
+    ONTYPE_FORMATTING("textDocument/onTypeFormatting"),
+    DEFINITION("textDocument/definition"),
+    CODE_ACTION("textDocument/codeAction"),
+    CODE_LENS("textDocument/codeLens"),
+    DOCUMENT_LINK("textDocument/documentLink"),
+    RENAME("textDocument/rename");
 
     private final String name;
 
@@ -39,9 +49,8 @@ public enum DynamicRegistrationMethods {
         this.name = name;
     }
 
-    public static DynamicRegistrationMethods forName(final String name) {
-        return Arrays.stream(DynamicRegistrationMethods.values()).filter(n -> n.name.equals(name))
-                .collect(Collectors.toList()).get(0);
+    public static Optional<DynamicRegistrationMethods> forName(final String name) {
+        return Arrays.stream(DynamicRegistrationMethods.values()).filter(n -> n.name.equals(name)).findAny();
     }
 
     public String getName() {

